### PR TITLE
Documentation: add ReviewingGuidelines

### PR DESCRIPTION
**Pull request by vdye**
This patch follows up on a discussion a few weeks ago in the Git IRC standup [1], where it was mentioned that it would be nice to have consistent definitions for common review terminology (like 'nit:'). The "ReviewingGuidelines" document created here builds on that idea, as well as past discussions around the idea of advice for reviewers (similar to the guidelines for new contributors in MyFirstContribution [2]).

The goal of this document is to clarify & standardize some of the more niche concepts important to the Git project ("What's cooking" emails, terminology), as well as provide general reviewing advice based on my observations of effective reviews from others on the mailing list. 

One thing that's particularly important to me here is that the advice presented here does *not* gatekeep or otherwise denigrate the personal preferences or style of reviewers. With that in mind, one of the things I'm looking for in reviews of this document is making sure that the tone & content reflect that more positive/encouraging intent. And, of course, I'm happy to hear what other tips & terminology people think would be helpful to include!

# Changes since V1
- Reorganized "Principles" section advice into "High-level guidance" and "Performing your review" subsections.
- Dropped recommendation to comment on cover letter with "LGTM" if you have no other recommendations (somewhat redundant with the "Completing a review" section, and such comments don't tend to add value except when coming from highly-experienced reviewers anyway).
- Added clarity & modified reasoning for why reviewing CC'd patches is helpful.
- Miscellaneous other revisions recommended by [3].

Thanks!
- Victoria

[1] https://colabti.org/irclogger/irclogger_log/git-devel?date=2022-08-29#l53
[2] https://git-scm.com/docs/MyFirstContribution
[3] https://lore.kernel.org/git/xmqqwnacibbm.fsf@gitster.g/

CC: derrickstolee@github.com
CC: Johannes.Schindelin@gmx.de
CC: steadmon@google.com
CC: chooglen@google.com
CC: gitster@pobox.com
cc: Elijah Newren <newren@gmail.com>
cc: Konstantin Ryabitsev <konstantin@linuxfoundation.org>

### CODE REVIEW

File: Documentation/ReviewingGuidelines.txt [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  related to - but outside the scope of - a series ("nice-to-have"s), or when
+  they represent only stylistic differences between the author and reviewer.
+
+- If you cannot complete a full review of a series all at once, consider letting
+  the author know (on- or off-list) if/when you plan to review the rest of the
+  series.
+
+- If you read and review a series but find nothing that warrants inline
+  commentary, reply to the series' cover letter to indicate that you've reviewed
+  the changes.
```
Comment (dscho) [inline Documentation/ReviewingGuidelines.txt:75]: I would like to suggest to add a few more bullet points:

- In many cases, the patches cannot be reviewed properly merely by looking at the diffs. For example, sometimes the default of three context lines in each diff hunk are insufficient to fully understand the intention of the patch. In such cases, applying those patches to a worktree and analyzing the full code (e.g. analyzing related parts of the code that are not shown in the diff, looking at the history of the code touched by the diff, building and running automated and manual tests) will help you provide much more helpful reviews.
- Often patches touch on areas that have been discussed before on the Git mailing list. It is helpful in such cases to find the link to the Git mailing list archive and most importantly to summarize the outcome(s) of the previous conversation.
- Reviews are most helpful when they are constructive, i.e. when they are designed to make the patch(es) better. For example, while it is rarely helpful to offer personal opinions, contributions often benefit from reviewers' insights about avoidable coding patterns and how to adjust the code according to best-known practices. Especially new contributors benefit from reviews that help adjust the code to Git's customs (preferably with an explanation why those customs exist).
- Be kind and graceful in your reviews. Keep in mind that the recipient of your review is a human being, not a machine. Ask only as much effort of the contributors as you are prepared to give. Lead by example, e.g. show the very thoroughness and diligence in your review that you would like to see in the contribution. If you ask the contributor to "do their homework", you can't just say "I haven't fully read the patch series yet but [...]" in the same mail.

File: Documentation/ReviewingGuidelines.txt [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+~~~~~~~~~~~~~~~~~~~
+Once each patch of a series is reviewed, the author (and/or other contributors)
+may discuss the review(s). This may result in no changes being applied, or the
+author will send a new version of their patch(es).
+
+After a series is rerolled in response to your or others' review, make sure to
+re-review the updates. If you are happy with the state of the patch series,
+explicitly indicate your approval (typically with a reply to the latest
+version's cover letter). Optionally, you can let the author know that they can
+add a "Reviewed-by: <you>" trailer to subsequent versions of their series.
```
Comment (dscho) [inline Documentation/ReviewingGuidelines.txt:87]: In some instances, the "What's cooking" mail explicitly asks whether a patch (series) is ready to advance to the next level, in which case a reply to that mail by someone familiar with the state of the review(s) is much appreciated, usually by a reviewer (self-assessments by the author do not carry as much weight here as the assessment of any reviewer involved in the mail thread). It is customary to change prefix such a mail's subject with "<branch-name>, (was Re: [...])", see e.g. [this mail](https://lore.kernel.org/git/a467d972-85a6-f4b0-7ec7-184add308e25@github.com/).

File: Documentation/ReviewingGuidelines.txt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+Terminology
+-----------
+nit: ::
+	Denotes a small issue that should be fixed, such as a typographical
+	error or mis-alignment of conditions in an `if()` statement.
```
Comment (dscho) [inline Documentation/ReviewingGuidelines.txt:93]: Maybe mention that nits are most helpful in the phase where the patch gets its final polish, so as not to distract from the more important aspects of the review.

File: Documentation/ReviewingGuidelines.txt [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+non-blocking: ::
+	Indicates to the reader that the following comment should not block the
+	acceptance of the patch or series. These are typically recommendations
+	related to code organization & style, or musings about topics related to
+	the patch in question, but beyond its scope.
+
+s/<before>/<after>::
+	Shorthand for "you wrote <before>, but I think you meant <after>,"
+	usually for misspellings or other typographical errors. The syntax is a
+	reference to the `vi` "Substitute" command.
```
Comment (dscho) [inline Documentation/ReviewingGuidelines.txt:106]: I believe that the `s/<regex>/<replacement>/` syntax originates from the [`ed`](https://en.wikipedia.org/wiki/Ed_(text_editor)) command and is "only" imitated by [`sed`](https://en.wikipedia.org/wiki/Sed), [`ex`](https://en.wikipedia.org/wiki/Ex_(text_editor)) (which `vi` uses when you enter the `:` mode), `vim` and `perl`.

Maybe reword it to something like "The syntax refers to the "substitute" command known from Unix tools like `ed`, `sed`, `vim` and `perl`, to silence nit-pickers such as myself?

File: Documentation/ReviewingGuidelines.txt [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+cover letter::
+	The "Patch 0" of a multi-patch series. This email describes the
+	high-level intent and structure of the patch series to readers on the
+	Git mailing list. It is also where the changelog notes and range-diff of
+	subsequent versions are provided by the author.
```
Comment (dscho) [inline Documentation/ReviewingGuidelines.txt:112]: It would be good to mention (probably in a second paragraph under the `cover letter` bullet point) that the preference for single-patch contributions is to _not_ have a separate mail with the cover letter, but to include a "cover letter" in the same mail as the patch, between the `---` line and the diffstat ([example](https://lore.kernel.org/git/de06f80c-4778-7e43-3a88-d7aeae788e0d@gmail.com/)).

**Review by dscho — APPROVED** [review]
What a great piece of work! Thank you so much for doing this, as you could probably guess, I care deeply about improving the reviewing situation on the Git mailing list.

I don't know whether you want to extend the commit message to give a little more context? I could imagine that it would be helpful to provide some color, e.g. mentioning that there have been some mismatches of expectations leading to friction lately and that [the idea of reviewers' guidelines has been raised in the Git standup on IRC](https://colabti.org/irclogger/irclogger_log/git-devel?date=2022-08-29#l53).

File: Documentation/ReviewingGuidelines.txt [hunk 6]
```diff
@@ -0,0 +1,116 @@
+Reviewing Patches in the Git Project
+====================================
+
+Introduction
+------------
+The Git development community is a widely distributed, diverse, ever-changing
+group of individuals. Asynchronous communication via the Git mailing list poses
+unique challenges in clarity when reviewing or discussing patches. This document
```
Comment (derrickstolee) [inline Documentation/ReviewingGuidelines.txt:8]:
```suggestion
unique challenges when reviewing or discussing patches. This document
```

File: Documentation/ReviewingGuidelines.txt (hunk 4 above)
Comment (derrickstolee) [reply Documentation/ReviewingGuidelines.txt:106]: My nit is that it should be `s/<before>/after/`, I think. ;)

File: Documentation/ReviewingGuidelines.txt (hunk 1 above)
Comment (derrickstolee) [reply Documentation/ReviewingGuidelines.txt:75]: More ideas to consider:

* To review patches locally in your worktree, you can apply the patches to your repository or wait for the branch to be tracked by the maintainer and check out that version.

* Some patches that perform significant refactoring can result in hard-to-read patches that become easier to read as commit diffs using `--color-moved` or `--ignore-space-change`.

* It is important to consider whether the changed code is covered by existing tests, or if a new test is required to exercise that behavior. With a local checkout of the patches, you can mutate the lines to be obviously wrong and see if that triggers a test failure. If you see a lack of testing, then recommend methods to test that behavior, including by writing a test yourself.

**Review by derrickstolee — APPROVED** [review]
Looking good! I only have a few recommendations to add. You covered the main ones I was thinking about.
